2.MLP_Multi_Classifier/MLP_utils/utils.py

```diff
@@ -0,0 +1,938 @@
+    # split train data into train-validate
+    X_train, X_test, Y_train, Y_test = train_test_split(
+        X_train, Y_train, test_size=test_proportion, random_state=seed, stratify=Y_train
+    )
+
+    # reset the index to avoid downstream errors
+    X_train = X_train.reset_index(drop=True)
+    Y_train = Y_train.reset_index(drop=True)
+    X_val = X_val.reset_index(drop=True)
+    Y_val = Y_val.reset_index(drop=True)
+    X_test = X_test.reset_index(drop=True)
+    Y_test = Y_test.reset_index(drop=True)
+
+    return X_train, X_test, X_val, Y_train, Y_test, Y_val
+
+
+# Class for x and y data
+class Dataset_formatter:
+    """
+    A class for formatting data for a data loader
+
+    Attributes:
+    ----------
+    X : Pandas DataFrame
+        the X dimension of data (features)
+    Y : Pandas DataFrame
+        the Y dimension of data (predictor)
+
+    Methods
+    -------
+    __len__:
+        returns the length of the X dimension
+    -------
+    __getitem__:
+        returns a row of the X and Y dimension given an index
+    """
+
+    def __init__(
+        self,
+        X,
+        Y,
+    ):
+        self.X = X
+        self.Y = Y
+
+    def __len__(self):
+        return len(self.X)
+
+    def __getitem__(self, idx):
+        return self.X[idx], self.Y[idx]
+
+
+# based on https://www.kaggle.com/code/ludovicocuoghi/pytorch-pytorch-lightning-w-optuna-opt
+def build_model_custom(
+    trial: optuna.study, in_features: int, final_layer_out_features: int, params: object
+) -> torch.nn.Sequential:
+    """Generate a flexible pytorch Neural Network Model that allows for
+    optuna hyperparameter optimization
+
+    Parameters
+    ----------
+    trial : optuna.study
+        an iteration of the optuna study object
+    in_features : int
+        the number of in features for the initial network layer
+    final_layer_out_features : int
+        the number of out features for the network final layer
+    params : object
+        parameter dataclass object to import constants and params
+
+    Returns
+    -------
+    torch.nn.Sequential
+        this returns in a dict the architecture of the model with optimized parameters
+    """
+
+    # number of hidden layers
+    # suggest.int takes into account the defined search space
+    n_layers = trial.suggest_int("n_layers", params.MIN_LAYERS, params.MAX_LAYERS)
+
+    #  layers will be added to this list and called upon later
+    layers = []
+
+    for i in range(n_layers):
+
+        # the number of units within a hidden layer
+        out_features = trial.suggest_int(
+            "n_units_l{}".format(i), params.LAYER_UNITS_MIN, params.LAYER_UNITS_MAX
+        )
+
+        layers.append(nn.Linear(in_features, out_features))
+        # activation function
+        layers.append(nn.ReLU())
+        # dropout rate
+        p = trial.suggest_float(
+            ("dropout_{}".format(i)), params.DROPOUT_MIN, params.DROPOUT_MAX
```

Comment:
Consider using an f-string here instead of `str.format()`.

Reply:
fixed! :D